cluster_experiments/power_analysis.py

```diff
@@ -524,3 +524,65 @@ def check_inputs(self):
         self.check_target_col()
         self.check_treatment()
         self.check_clusters()
+
+
+class PowerAnalysisWithPreExperimentData(PowerAnalysis):
+    """
+    Same as PowerAnalysis, but adding pre-experiment param during split and perturbation. The PowerAnalysis class
+    accepts a pre_experiment param, however this is only used for cuped purposes (where we add a column to df). This
+    child class is used for cases where the pre experiment df is also available when the class is instantiated
+    """
+
+    def split(self, df: pd.DataFrame) -> pd.DataFrame:
+        treatment_df = self.splitter.assign_treatment_df(df)
+        self.log_nulls(treatment_df)
+
+        treatment_df = treatment_df.query(
+            f"{self.treatment_col}.notnull()", engine="python"
+        ).query(
+            f"{self.treatment_col}.isin(['{self.treatment}', '{self.control}'])",
+            engine="python",
+        )
+
+        return treatment_df
+
+    def perturbate(
+        self, treatment_df: pd.DataFrame, average_effect: Optional[float]
+    ) -> pd.DataFrame:
+
+        pre_experiment_df = treatment_df.query(
+            f"{self.analysis.time_col} <= '{self.analysis.intervention_date}'"
+        )
+
+        treatment_df = treatment_df[
+            (treatment_df[self.analysis.time_col] > self.analysis.intervention_date)
+        ]
+
+        perturbed_df = self.perturbator.perturbate(
+            treatment_df, average_effect=average_effect
+        )
+
+        # get treatment cluster
+        treatment_cluster = perturbed_df.loc[
```

Comment:
this method is a bit long, I'd try to:
* move part of the logic in here to another method
* unit-test it. this has quite a bit of logic and it's a good candidate to test that it's doing what is expected